lms/djangoapps/completion/models.py

```diff
@@ -106,6 +106,81 @@ def submit_completion(self, user, course_key, block_key, completion):
         return obj, isnew
 
 
+class AggregateCompletionManager(models.Manager):
+    """
+    Custom manager for AggregateCompletion model.
+
+    """
+
+    def submit_completion(self, user, course_key, block_key, earned, possible):
+        """
+        Inserts and Updates the completion AggregateCompletion for the specified record.
+
+        Parameters:
+            * user (django.contrib.auth.models.User): The user for whom the
+              completion is being submitted.
+            * course_key (opaque_keys.edx.keys.CourseKey): The course in
+              which the submitted block is found.
+            * block_key (opaque_keys.edx.keys.UsageKey): The block that has had
+              its completion changed.
+            * earned: The fractional completion
+              value of the block (0.0 = incomplete, 1.0 = complete).
+            * possible: The fractional completion
+              value of the block (0.0 = incomplete, 1.0 = complete).
+
+        Return Value:
+            (BlockCompletion, bool): A tuple comprising the created or updated
+            BlockCompletion object and a boolean value indicating whether the
+            object was newly created by this call.
+
+        Raises:
+
+            ValueError:
+                If the wrong type is passed for one of the parameters.
+
+            django.core.exceptions.ValidationError:
+                If a float is passed that is not between 0.0 and 1.0.
+
+            django.db.DatabaseError:
+                If there was a problem getting, creating, or updating the
+                BlockCompletion record in the database.
+
+                This will also be a more specific error, as described here:
+                https://docs.djangoproject.com/en/1.11/ref/exceptions/#database-exceptions.
+                IntegrityError and OperationalError are relatively common
+                subclasses.
+        """
+
+        # Raise ValueError to match normal django semantics for wrong type of field.
+        if not isinstance(course_key, CourseKey):
+            raise ValueError(
+                "course_key must be an instance of `opaque_keys.edx.keys.CourseKey`.  Got {}".format(type(course_key))
+            )
+
+        percent = earned / possible
+
+        if waffle.waffle().is_enabled(waffle.ENABLE_COMPLETION_TRACKING):
+            obj, isNew = self.get_or_create(
+                user=user,
+                course_key=course_key,
+                aggregation_name="??",
```

Comment:
This should be a parameter. In the initial implementation, it describes at what level this aggregation is happening - i.e. it's either `course`, `chapter`, or `sequential`, or `vertical` (course, section, subsection, or unit).

Reply:
Do you think it makes sense to make aggregation_name a ChoiceField instead of a CharField? This has discrete values that may or may not have more values in the future?